src/tests/src/solver/expressions/test_CompareVisitor.cpp

```diff
@@ -0,0 +1,123 @@
+ * along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+ */
+
+#define WIN32_LEAN_AND_MEAN
+
+#include <boost/test/unit_test.hpp>
+
+#include <antares/solver/expressions/Registry.hxx>
+#include <antares/solver/expressions/nodes/ExpressionsNodes.h>
+#include <antares/solver/expressions/visitors/CloneVisitor.h>
+#include <antares/solver/expressions/visitors/CompareVisitor.h>
+#include <antares/solver/expressions/visitors/PrintVisitor.h>
+
+using namespace Antares::Solver;
+using namespace Antares::Solver::Nodes;
+using namespace Antares::Solver::Visitors;
+
+// Only necessary for BOOST_CHECK_EQUAL
+
+BOOST_AUTO_TEST_SUITE(_CompareVisitor_)
+
+static Node* createSimpleExpression(Registry<Node>& registry, double param)
+{
+    Node* var1 = registry.create<LiteralNode>(param);
+    Node* param1 = registry.create<ParameterNode>("param1");
+    Node* expr = registry.create<AddNode>(var1, param1);
+    return expr;
+}
+
+BOOST_FIXTURE_TEST_CASE(comparison_to_self_simple, Registry<Node>)
+{
+    CompareVisitor cmp;
+    Node* expr = createSimpleExpression(*this, 65.);
+    BOOST_CHECK(cmp.dispatch(*expr, *expr));
+}
+
+BOOST_FIXTURE_TEST_CASE(comparison_to_other_same, Registry<Node>)
+{
+    CompareVisitor cmp;
+    auto create = [this] { return createSimpleExpression(*this, 65.); };
+    Node* expr1 = create();
+    Node* expr2 = create();
+    BOOST_CHECK(cmp.dispatch(*expr1, *expr2));
+}
+
+BOOST_FIXTURE_TEST_CASE(comparison_to_other_different, Registry<Node>)
+{
+    CompareVisitor cmp;
+    Node* expr1 = createSimpleExpression(*this, 64.);
+    Node* expr2 = createSimpleExpression(*this, 65.);
+    BOOST_CHECK(!cmp.dispatch(*expr1, *expr2));
+}
+
+static Node* createComplexExpression(Registry<Node>& registry)
+{
+    // NOTE : this expression makes no sense, only for testing purposes
+    // NOTE2 : Some elements are re-used (e.g simple), this is valid since memory is handled
+    // separately (no double free)
+
+    Node* simple = createSimpleExpression(registry, 42.);
+    Node* neg = registry.create<NegationNode>(simple);
+    Node* mult = registry.create<MultiplicationNode>(simple, neg);
+    Node* comp = registry.create<ComponentParameterNode>("hello", "world");
+    Node* div = registry.create<DivisionNode>(mult, comp);
+    Node* div2 = registry.create<DivisionNode>(div, simple);
+    Node* add = registry.create<AddNode>(div, div2);
+    Node* sub = registry.create<SubtractionNode>(add, neg);
+    Node* cmp = registry.create<GreaterThanOrEqualNode>(sub, add);
+    Node* pf = registry.create<PortFieldNode>("port", "field");
+    Node* addf = registry.create<AddNode>(pf, cmp);
+    return addf;
+}
+
+BOOST_FIXTURE_TEST_CASE(comparison_to_self_complex, Registry<Node>)
+{
+    CompareVisitor cmp;
+    Node* expr = createComplexExpression(*this);
+    BOOST_CHECK(cmp.dispatch(*expr, *expr));
+}
+
+BOOST_FIXTURE_TEST_CASE(comparison_to_clone_complex, Registry<Node>)
```

Comment:
Here is something I don't understand : I suppose that, in real life, we should associate one registry to one expression / tree. And it would make sense : a clone life time should be independent from the life time of the tree it was created from.
Here we have one registry for a tree and its clone. Is it intended ?

Reply:
If the previous remark is relevant, something else can be pointed : 
if we must associate one registry to one tree / expression, can we have 2 independent objects alive for an expression (I mean the registry and the expression itself) ?
User of an expression must handle 2 objects, which are mutually cohesive : one cannot exist without the other.
So if we think that registry class is useful, we should find a way the 'link' more closely expressions and registry. For example having a class **Expression** or **Tree** or **AST** that contain a root node and a registry as data member.

We still have the solution where node are pointed to with **std::shared_ptr** (which we know to be flawed).

Reply:
the notion of tree is not defined here, we only care about nodes.

Reply:
I rephrase my question : 
here we have only have one registry for 2 expressions (or tree) : **expr** and **expr_cloned**.
Is it normal ?
Should not we have 2 registries : one associated to each expressions ?